### PR TITLE
Add RestChunkedToXContentListener version that releases transport responses

### PR DESCRIPTION
**Pull request by original-brownbear**
This is needed as part of #102030, we need to be able to release search/multi-search and maybe other transport messages after they've been serialized as REST responses.
Just doing this in a separate PR to keep things easily reviewable, obviously this is without effect at this point.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/action/RestChunkedToXContentListener.java [hunk 1]
```diff
@@ -37,10 +38,17 @@ public RestChunkedToXContentListener(RestChannel channel, ToXContent.Params para
...
         channel.sendResponse(
-            RestResponse.chunked(getRestStatus(response), ChunkedRestResponseBody.fromXContent(response, params, channel, null))
+            RestResponse.chunked(
+                getRestStatus(response),
+                ChunkedRestResponseBody.fromXContent(response, params, channel, releasableFromResponse(response))
+            )
         );
     }
 
+    protected Releasable releasableFromResponse(Response response) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/rest/action/RestChunkedToXContentListener.java:48]: Not in love with this approach, but I didn't want to duplicate this across both rest actions and I expect us to need this elsewhere as well eventually.

File: server/src/main/java/org/elasticsearch/rest/action/RestRefCountedChunkedToXContentListener.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ */
+public class RestRefCountedChunkedToXContentListener<Response extends ChunkedToXContent & RefCounted> extends RestChunkedToXContentListener<
+    Response> {
+    public RestRefCountedChunkedToXContentListener(RestChannel channel) {
+        super(channel);
+    }
+
+    @Override
+    protected Releasable releasableFromResponse(Response response) {
+        return response::decRef;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/rest/action/RestRefCountedChunkedToXContentListener.java:28]: Maybe wrap in `Releasables#assertOnce` for paranoia?

File: server/src/main/java/org/elasticsearch/rest/action/RestChunkedToXContentListener.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/rest/action/RestChunkedToXContentListener.java:48]: Eh seems ok to me. The "proper" approach would be to make `ChunkedToXContent extends RefCounted` too and then add an `incRef`/`decRef` pair here, but that seems excessive to me for now.

**Review by DaveCTurner — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/rest/action/RestRefCountedChunkedToXContentListener.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/rest/action/RestRefCountedChunkedToXContentListener.java]: ++ big fan of paranoia => done, also this actually needed a preceding `incRef` now that the other changes are in, we must be ref-count neutral like any other listener here end-to-end :)

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**